### PR TITLE
pluma-view: sync the font in context-menu with gsettings key

### PR DESCRIPTION
**Pull request by sc0w**

### CODE REVIEW

File: pluma/pluma-view.c [hunk 1]
```diff
@@ -699,13 +714,27 @@ pluma_override_font (const gchar          *item,
...
 	else
 		css = g_strdup_printf ("%s { %s %s %s %s }", item, family, weight, style, size);
 
 	gtk_css_provider_load_from_data (provider, css, -1, NULL);
 
-	if (!provider_added) {
+	if (!provider_added)
+	{
+		GSettings *settings;
+		settings = g_settings_new ("org.mate.interface");
```
Comment (muktupavels) [inline pluma/pluma-view.c]: You are leaking `GSettings` object, no?

File: pluma/pluma-view.c [hunk 2]
```diff
@@ -699,13 +714,27 @@ pluma_override_font (const gchar          *item,
 	prov_str = gtk_css_provider_to_string (provider);
 
 	if (g_str_has_prefix (prov_str, "textview") && g_str_has_prefix (item, ".context-menu"))
+	{
+		if (strstr (prov_str, ".context-menu"))
+		{
+			prov_str = g_strdelimit (prov_str, "}", '\0');
+			prov_str = g_strdup_printf ("%s}", prov_str);
```
[the remark this answers is not in the document]
Comment (lukefromdc) [reply pluma/pluma-view.c]: The last line in pluma_override_font with this PR is

g_free (prov_str);

[the remark this answers is not in the document]
Comment (muktupavels) [reply pluma/pluma-view.c]: And?

**Comment by lukefromdc** [comment]
Also note that gtk_css_provider_get_default() has just been deprecated in GTK 3.23 and we should replace it. A static variable will work and hold its value for the lifetime of a Pluma run, that's really all  gtk_css_provider_get_default() does anymore..

File: pluma/pluma-view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukefromdc) [reply pluma/pluma-view.c]: What do you mean? I did not grow up with text message shorthand and do not understand it

[the remark this answers is not in the document]
Comment (lukefromdc) [reply pluma/pluma-view.c]: That block only sets one new variable, which is freed later

[the remark this answers is not in the document]
Comment (muktupavels) [reply pluma/pluma-view.c]: What did not you understand? There is memory leak...

That new variable is newly-allocated variable so you free that, yes. But who frees original variable?

**Comment by lukefromdc** [comment]
https://github.com/mate-desktop/pluma/pull/345/commits/e5e7b64a070c82d5e9f6ae2264dd4bf31b2d6030
replaces newly deprecated gtk_css_provider_get_default() with  a static provider and also uses a static settings variable. If that is freed it no longer works, as it needs to be available as long as Pluma is running to track user changes to the font.

File: pluma/pluma-view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukefromdc) [reply pluma/pluma-view.c]: You are not giving me enough information to figure it out, so I will leave that to the others for now.

**Comment by sc0w** [comment]
@lukefromdc I think `static` in `Gsettings` isn't needed, the signal works fine whithout it, but I have no problem if you decide to do that.

**Comment by lukefromdc** [comment]
Thanks. I made the settings static because doing so stopped a potential memory leak. If I freed settings fonts were no longer applied. We need to keep settings not for the life of the function, but the life of the Pluma run, and that is the job of a static variable.

File: pluma/pluma-view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukefromdc) [reply pluma/pluma-view.c]: What happens when a variable is loaded with g_strdup_printf of its own previous value(is this to change the format?)? I don't know enough offhand about the string functions to know what is done with the previous value.

[the remark this answers is not in the document]
Comment (muktupavels) [reply pluma/pluma-view.c]: Maybe time to learn?

[the remark this answers is not in the document]
Comment (lukefromdc) [reply pluma/pluma-view.c]: If you are going to criticize instead of help, we are going to have to ignore the criticism and proceed to merge this.

[the remark this answers is not in the document]
Comment (muktupavels) [reply pluma/pluma-view.c]: Sorry, I already helped you by telling that you have memory leak. I am not going to fix your mistakes. If you think that all is good, merge it.

[the remark this answers is not in the document]
Comment (lukefromdc) [reply pluma/pluma-view.c]: I will leave this for the others, maybe someone else on the team can figure it out

[the remark this answers is not in the document]
Comment (raveit65) [reply pluma/pluma-view.c]: Why not telling how to fix the leak?
Don't you see that nobody can satisfied your concerns?
.... to much mails about in my mailbox -_-

[the remark this answers is not in the document]
Comment (vkareh) [reply pluma/pluma-view.c]: I think the issue might be that we're reusing the variable, so we end up re-allocating memory every time you assign to it, and only free it once at the very end. The intermediary steps would get lost, and thus leak. Assuming this is the case (could you confirm if that is true @muktupavels?), we should change this bit to either not reuse the variable (i.e. doing the entire string assignment in a single expression), or have a temp variable that gets freed after every new allocation.

**Comment by sc0w** [comment]
I am not sure if there is a memory leak, but I did some changes.

In my tests, all is working like before

Note: `g_strdelimit` doesn't need to be freed

**Comment by lukefromdc** [comment]
I can confirm this latest change works as before

File: pluma/pluma-view.c [hunk 3]
```diff
@@ -694,18 +709,36 @@ pluma_override_font (const gchar          *item,
...
 	gtk_css_provider_load_from_data (provider, css, -1, NULL);
 
-	if (!provider_added) {
+	if (!provider_added)
+	{
+		static GSettings *settings; /*We need this for the life of the provider and program*/
+		settings = g_settings_new ("org.mate.interface");
+		g_signal_connect (settings,
+				  "changed::" "font-name",
+				  G_CALLBACK (contextmenu_font_changed_cb), NULL);
```
Comment (vkareh) [inline pluma/pluma-view.c]: Here you can probably pass `widget` instead of `NULL`, that way you get the right object in the callback when doing `GTK_WIDGET (user_data)`.

File: pluma/pluma-view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (muktupavels) [reply pluma/pluma-view.c]: @vkareh yes.

**Comment by sc0w** [comment]
The widget only is needed the first time the function `pluma_override_font` is called

I added this change to the latest commit:

```patch
@@ -675,7 +675,7 @@ contextmenu_font_changed_cb (GSettings *settings,
 	sys_font_desc = get_system_font ();
 	if (sys_font_desc)
 	{
-		pluma_override_font (".context-menu", GTK_WIDGET (user_data), sys_font_desc);
+		pluma_override_font (".context-menu", NULL, sys_font_desc);
 		pango_font_description_free (sys_font_desc);
 	}
 }
```

**Comment by raveit65** [comment]
Is this ready for testing?

**Comment by sc0w** [comment]
is ready for me

**Comment by sc0w** [comment]
@muktupavels 

In your proposed changes you changed everything, that is another history, and for me it must be discussed and tested in another PR

Anyway I prefer to do the things easily, with minimal changes

**Comment by muktupavels** [comment]
What is everything? The change is quite simple and probably could be split in few commits for better history.

For example renaming from `pluma_prefs_manager_get_system_font` to `pluma_prefs_manager_get_monospace_font_name` can be done as separate commit... But that part is not really needed, you can keep function unchanged and add new function for `font-name`...

I don't find your proposal easy... For example, you did not answer why you need to create new GSettings object when there is already setting object for that schema? Adding new function to preferences to get that font name is way better thing then creating new object and is not harder...

Also `get_system_font` makes me ask - why do you need that? There is already `pluma_prefs_manager_get_system_font`. And both use interface schema. So why duplicate code? Of course in this case they mean different things, but it is hard to tell that from function names.

Also just noticed that you actually create two new GSettings objects so in total there is even 3 object for one schema... You are just wasting resources.

Same with `gtk_css_provider_to_string`, check what does that function:
https://gitlab.gnome.org/GNOME/gtk/blob/master/gtk/gtkcssprovider.c#L1925

Hint: it does not return string you previously set. It has to rebuild that string.

It might be easier to you, but that is definitely not better for pluma.

Also one more idea. If style provider would be added to `PlumaView` widget not screen then maybe you would not need to change/update context-menu font. From:
https://developer.gnome.org/gtk3/stable/GtkStyleContext.html#gtk-style-context-add-provider

> Note that a style provider added by this function only affects the style of the widget to which context belongs.

So adding provider to widget might be another solution.

And actually there might be even simpler thing. Have you tried to set font for `textview text` not `textview`? According to gtk inspector .context-menu is child of textview so it might inherit css from it, but if you do `textview text { ... }` then it should not affect context-menu...

In short you should look for best solution not the one that easier to you.

**Comment by muktupavels** [comment]
P.S. `textview text` in quick test looks like does not work... I can change color, bet font-* are ignored.

**Comment by sc0w** [comment]
You say words without tests... You proposed code without tests... You review PRs without tests...

I am not interested in working more in this PR.

(Of course, if someone makes new changes, I will test it)

**Comment by muktupavels** [comment]
Well it might be bug in GTK if `font-*` things does not work with `textview text {}`...

**Comment by lukefromdc** [comment]
One point about Albert: He is probably the coder here who knows the most about GTK. He can't test code for MATE directly because he does not have a MATE build environment set up. He develops the very similar gnome-flashback session that shares a lot of code. There is a lot to be said for co-development of these very similar environments and codebases.

If an issue we are working on has already been fixed in gnome-flashback, it makes sense to use the same fix. Conversely, if we fix it first Albert can base fixes for gnome-flashback on those fixes. Reducing code divergence between these forks means bugs are more likely to be common to both and this fixable in both at once.  New commits for one become more easily ported to the other.

That said, we don't have a big enough team to turn every PR into an overall refactoring effort. Doing so (with enough people and time to make it work) would have many long term benefits, but with such a small team as we have and no distro offering MATE development releases in their rolling release, new bugs and regressions could quickly spoil the party.

This is not like when Ubuntu alphas installed development versions of GNOME back in the GNOME 2 days or first few cycles of GNOME 3. Now Ubuntu and Debian (and probably a lot more distros) seem to ignore development releases even in their development versions, which takes out the majority of pre-release version testing.

At this point, we have a basically solid DE, but with lots of user issues open and the ever-changing new issues created by HiDPI displays, by new features in external applications, and the usual changes in underlying libs. We are still chasing a lot of deprecated functions from the GTK3 changeover, and new ones are still coming. Sometimes fixing these introduces new bugs, which then have to be fixed.

That being so, for a lot of this stuff we need to keep it simple, and for longer term work some things might need refactoring branches which can be merged when and only when they are really ready. This is how the Kdenlive devs have been handling a systematic refactoring of their entire codebase, with some portions now merged to released versions and others still under development.

**Comment by muktupavels** [comment]
Already in master:
- https://github.com/mate-desktop/pluma/commit/310bfc249aac3dd80511d3dade14aa06f3d9769a
- https://github.com/mate-desktop/pluma/commit/8f3b1d625c86cc57b51c8acb3b1ab4e5f59d4d4d
- https://github.com/mate-desktop/pluma/commit/e451bfe61952156ba6bdb63700f1089c2c52f766
- https://github.com/mate-desktop/pluma/commit/9fa364980a154864e7217b0160ac7d570cd76d79

And this is 5th merge request for same thing... You had bugs anyway and I would call that as "refactoring"... Why do work now to redo that later? Do the right thing from start... As example - is it that hard to add extra function to get `font-name` from existing GSettings object?

**Comment by raveit65** [comment]
For me @muktupavels role is complete unclear.
If someone doing reviews here he should also install MATE in a VM and test suggestion.
Every teacher in a school do practice test before he do suggestions.
Saying i'm gnome-(flashback) and i can't do that isn't sufficient.
I have also gnome in VMs running.
Sorry, that an authors of a PR stops his work , isn't acceptable.

Congratulations

**Comment by lukefromdc** [comment]
For my own work, I have to be able to build MATE, compiz, Nemo, Nautilus, gnome-panel, and Kdenlive. Setting up a build environment is not that complex and only has to be done once.

**Comment by raveit65** [comment]
I will merge it, but a later refactor might be a good idea regardless.
We should keep in mind a later refactor to clean up the static variable issue.
